include/extractor/pearce_scc.hpp

```diff
@@ -0,0 +1,154 @@
+#ifndef PEARCE_SCC_HPP
+#define PEARCE_SCC_HPP
+
+#include "extractor/node_based_edge.hpp"
+#include "extractor/query_node.hpp"
+#include "util/deallocating_vector.hpp"
+#include "util/percent.hpp"
+#include "util/typedefs.hpp"
+
+#include "util/integer_range.hpp"
+#include "util/log.hpp"
+#include "util/std_hash.hpp"
+#include "util/timing_util.hpp"
+
+#include "osrm/coordinate.hpp"
+#include <boost/assert.hpp>
+#include <cstdint>
+
+#include <algorithm>
+#include <climits>
+#include <memory>
+#include <stack>
+#include <vector>
+
+namespace osrm
+{
+namespace extractor
+{
+
+template <typename GraphT> class PearceSCC
+{
+    std::vector<NodeID> components_index;
+    std::vector<std::size_t> component_size_vector;
+    std::shared_ptr<const GraphT> m_graph;
+
+  public:
+    PearceSCC(std::shared_ptr<const GraphT> graph)
```

Comment:
I don't understand why we need to take shared ownership over a graph here.

A `const GraphT&` should be good enough? The callsite can unwrap if needs be.